pkg/nfd-gc/nfd-gc.go

```diff
@@ -172,7 +180,26 @@ func (n *nfdGarbageCollector) garbageCollect() {
 		klog.ErrorS(err, "failed to list NodeResourceTopology objects")
 	} else {
 		for _, nrt := range nrts.Items {
+			delete := false
+
 			if !nodeNames.Has(nrt.Name) {
+				delete = true
+			}
+
+			if label, ok := nrt.Labels[NRTOwnerPodLabel]; ok {
+				if s := strings.Split(string(label), "/"); len(s) == 2 {
+					ns := s[0]
+					pod := s[1]
+					_, err := n.k8sClient.CoreV1().Pods(ns).Get(context.TODO(), pod, metav1.GetOptions{})
```

Comment:
Hmm, I think there's a window for nasty race condition in case of (topology-updater) pod restarts:

a) topology-updater pod is deleted
b) nfd-gc kicks in and doesn't see the pod (even if it's temporarily gone)
c) nfd-gc GCs the NRT
d) -> problems 

One possible solution would be to have keep a list (e.g. `map[string]string`, key is the NRT name, value is the pod ns/name) of stale NRTs. On the first GC round, if the topology-updater pod is not found mark the NRT on this list. On the next GC round, if the pod is still not found (and the pod ns/name hasn't changed) delete the NRT. WDYT?

ping @PiotrProkop @ffromani

Reply:
I think we'll end up with this race condition every time when nfd-topology-updater will be upgraded in the cluster, as the name of the pod(created by daemonset) will change on the node.
Another idea would be to somehow check if `nfd-topology-updater` is being handled by some controller(via OwnerReference) and just check if this controller still exists? 
I am still debating if I like this idea or not as it may make the code much more complicated and I don't know if it's worth it 😄 even though 99% of the deployment should create `nfd-topology-updater` via daemon-sets we can't guarantee that, so we would need to also write the code for this 1% of deployments.

Reply:
With what I suggested the race should be solved. Don't GC immediately but wait for two GC periods. If the version (or pod template spec in the DaemonSet whatever) is updated the new pod will update the owner pod label -> the NRT would not be GC'd, even if the new pod would be (temporarily) gone, too. Only when the owner pod (marked by the special label, we could use annotation, too) is gone for two consecutive GC periods the NRT would be GC'd.

This method doesn't depend on how the topology-updater is deployed (daemonset, deployment, custom controller) as the "owner ref" label is set by the topology-updater pod.